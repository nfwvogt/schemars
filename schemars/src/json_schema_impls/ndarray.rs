use crate::gen::SchemaGenerator;
use crate::schema::*;
use crate::JsonSchema;
use ndarray::ArrayBase;
use ndarray::Dimension;
use ndarray::Data;
use ndarray::Dim;
use ndarray::Ix;
use ndarray::IxDyn;


impl <T> JsonSchema for Dim<T> where T: JsonSchema{
    fn schema_name() -> String {
        format!("Dimension_of_type_{}", <T>::schema_name())
    }

    fn json_schema(gen: &mut crate::gen::SchemaGenerator) -> Schema {
        <T>::json_schema(gen)
    }
}

impl JsonSchema for IxDyn {
    fn schema_name() -> String {
        "DynamicArrayDimension".to_owned()
    }

    fn json_schema(gen: &mut crate::gen::SchemaGenerator) -> Schema {
        Vec::<Ix>::json_schema(gen)
    }
}

impl<T, D, S> JsonSchema for ArrayBase<S, D>
where T: JsonSchema, D: Dimension + JsonSchema, S:Data<Elem=T> {
    fn schema_name() -> String {
        format!("Array_of_type_{}_{}",T::schema_name(), D::schema_name())
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        let mut schema = SchemaObject {
            instance_type: Some(InstanceType::Object.into()),
            ..Default::default()
        };
        let obj = schema.object();
        obj.required.insert("v".to_owned());
        obj.required.insert("dim".to_owned());
        obj.required.insert("data".to_owned());
        obj.properties
            .insert("v".to_owned(), <u8>::json_schema(gen));
        obj.properties
            .insert("dim".to_owned(), D::json_schema(gen));
        obj.properties
            .insert("data".to_owned(), <Vec<T>>::json_schema(gen));
        schema.into()
    }
}