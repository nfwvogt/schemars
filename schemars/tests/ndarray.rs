mod util;
use ndarray::{Array0, Array1, Array2, Array3, Array4};
use schemars::{schema_for, JsonSchema};
use jsonschema::{Draft, JSONSchema};
use util::*;
use serde::{Serialize, Deserialize};

#[derive(Debug, JsonSchema, Serialize, Deserialize)]
struct TestStruct{
    zero: Array0<f32>,
    one: Array1<f64>,
    two: Array2<u32>,
    three: Array3<u64>,
    four: Array4<i32>
}


#[test]
fn ndarray() -> TestResult {
    test_default_generated_schema::<TestStruct>("ndarray")
}

#[test]
fn array_serialisation() {
    let test_schema = schema_for!(TestStruct);
    let schema = serde_json::to_string(&test_schema).unwrap();
    let schema_value: serde_json::Value = serde_json::from_str(&schema).unwrap();
    let test_struct = TestStruct{zero: Array0::zeros(()), one: Array1::zeros(2), two: Array2::zeros((3,4)), three: Array3::zeros((1,3,4)), four: Array4::zeros((2,1,2,2))};
    let test_json = serde_json::to_string(&test_struct).expect("serialisation of test struct failed");
    let test_value: serde_json::Value = serde_json::from_str(&test_json).unwrap();
    let compiled_schema = JSONSchema::options().with_draft(Draft::Draft7).compile(&schema_value).expect("reading test_schema failed");
    let validation_result = compiled_schema.validate(&test_value);
    assert!(validation_result.is_ok());
}