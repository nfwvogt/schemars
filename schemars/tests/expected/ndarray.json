{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "TestStruct",
    "type": "object",
    "required": [
      "four",
      "one",
      "three",
      "two",
      "zero"
    ],
    "properties": {
      "four": {
        "$ref": "#/definitions/Array_of_type_int32_Dimension_of_type_Array_size_4_of_uint"
      },
      "one": {
        "$ref": "#/definitions/Array_of_type_double_Dimension_of_type_Array_size_1_of_uint"
      },
      "three": {
        "$ref": "#/definitions/Array_of_type_uint64_Dimension_of_type_Array_size_3_of_uint"
      },
      "two": {
        "$ref": "#/definitions/Array_of_type_uint32_Dimension_of_type_Array_size_2_of_uint"
      },
      "zero": {
        "$ref": "#/definitions/Array_of_type_float_Dimension_of_type_EmptyArray"
      }
    },
    "definitions": {
      "Array_of_type_double_Dimension_of_type_Array_size_1_of_uint": {
        "type": "object",
        "required": [
          "data",
          "dim",
          "v"
        ],
        "properties": {
          "data": {
            "type": "array",
            "items": {
              "type": "number",
              "format": "double"
            }
          },
          "dim": {
            "type": "array",
            "items": {
              "type": "integer",
              "format": "uint",
              "minimum": 0.0
            },
            "maxItems": 1,
            "minItems": 1
          },
          "v": {
            "type": "integer",
            "format": "uint8",
            "minimum": 0.0
          }
        }
      },
      "Array_of_type_float_Dimension_of_type_EmptyArray": {
        "type": "object",
        "required": [
          "data",
          "dim",
          "v"
        ],
        "properties": {
          "data": {
            "type": "array",
            "items": {
              "type": "number",
              "format": "float"
            }
          },
          "dim": {
            "type": "array",
            "maxItems": 0
          },
          "v": {
            "type": "integer",
            "format": "uint8",
            "minimum": 0.0
          }
        }
      },
      "Array_of_type_int32_Dimension_of_type_Array_size_4_of_uint": {
        "type": "object",
        "required": [
          "data",
          "dim",
          "v"
        ],
        "properties": {
          "data": {
            "type": "array",
            "items": {
              "type": "integer",
              "format": "int32"
            }
          },
          "dim": {
            "type": "array",
            "items": {
              "type": "integer",
              "format": "uint",
              "minimum": 0.0
            },
            "maxItems": 4,
            "minItems": 4
          },
          "v": {
            "type": "integer",
            "format": "uint8",
            "minimum": 0.0
          }
        }
      },
      "Array_of_type_uint32_Dimension_of_type_Array_size_2_of_uint": {
        "type": "object",
        "required": [
          "data",
          "dim",
          "v"
        ],
        "properties": {
          "data": {
            "type": "array",
            "items": {
              "type": "integer",
              "format": "uint32",
              "minimum": 0.0
            }
          },
          "dim": {
            "type": "array",
            "items": {
              "type": "integer",
              "format": "uint",
              "minimum": 0.0
            },
            "maxItems": 2,
            "minItems": 2
          },
          "v": {
            "type": "integer",
            "format": "uint8",
            "minimum": 0.0
          }
        }
      },
      "Array_of_type_uint64_Dimension_of_type_Array_size_3_of_uint": {
        "type": "object",
        "required": [
          "data",
          "dim",
          "v"
        ],
        "properties": {
          "data": {
            "type": "array",
            "items": {
              "type": "integer",
              "format": "uint64",
              "minimum": 0.0
            }
          },
          "dim": {
            "type": "array",
            "items": {
              "type": "integer",
              "format": "uint",
              "minimum": 0.0
            },
            "maxItems": 3,
            "minItems": 3
          },
          "v": {
            "type": "integer",
            "format": "uint8",
            "minimum": 0.0
          }
        }
      }
    }
  }